nucliadb_node/src/merge/work.rs

```diff
@@ -0,0 +1,142 @@
+// Copyright (C) 2021 Bosutech XXI S.L.
+//
+// nucliadb is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at info@nuclia.com.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+//
+
+use std::collections::{HashMap, VecDeque};
+
+use itertools::Itertools;
+
+use super::request::MergePriority;
+use super::request::MERGE_PRIORITIES_COUNT;
+
+/// Work queue structure for merge scheduler. It serves as a priority queue with
+/// deduplication of elements
+pub struct WorkQueue<T> {
+    queues: HashMap<MergePriority, VecDeque<T>>,
+}
+
+// `WorkQueue` implementation uses a hash map to decouple the code from
+// `MergePriority`
+//
+// As the amount of priorities is small, we can consider hash map operations
+// take constant time. Using a hash map decouples the code from hardcoding
+// priorities in our methods.
+impl<T: PartialEq> WorkQueue<T> {
+    pub fn new() -> Self {
+        let mut queues = HashMap::with_capacity(MERGE_PRIORITIES_COUNT);
+        queues.insert(MergePriority::WhenFree, VecDeque::new());
+        queues.insert(MergePriority::Low, VecDeque::new());
+        queues.insert(MergePriority::High, VecDeque::new());
+
+        Self {
+            queues,
+        }
+    }
+
+    /// Pushes an item to the queue if it wasn't there yet with an specific
+    /// priority. Deduplication only occurrs in the requested priority
+    pub fn push(&mut self, item: T, priority: MergePriority) {
+        let queue = self.queues.get_mut(&priority).expect("Priority queue must exist");
+
+        if !queue.contains(&item) {
+            queue.push_back(item);
+        }
+    }
+
+    /// Removes the greatest item from the queue and returns it, or `None` if
+    /// it's empty
+    pub fn pop(&mut self) -> Option<T> {
+        for (_, queue) in self.queues.iter_mut().sorted_by_key(|x| x.0).rev() {
+            if let Some(item) = queue.pop_front() {
+                return Some(item);
+            }
+        }
+        None
+    }
+
+    pub fn len(&self) -> usize {
+        self.queues.values().fold(0, |acc, queue| acc + queue.len())
```

Comment:
```suggestion
        self.queues.values().reduce(|acc, queue| acc + queue.len())
```

Nitpick: I like using reduce when the initializer is the default, I find it a bit more explicit.